dbms/src/Flash/Mpp/ReceivedMessageQueue.cpp

```diff
@@ -0,0 +1,160 @@
+{
+void injectFailPointReceiverPushFail(bool & push_succeed [[maybe_unused]], ReceiverMode mode)
+{
+    switch (mode)
+    {
+    case ReceiverMode::Local:
+        fiu_do_on(FailPoints::random_receiver_local_msg_push_failure_failpoint, push_succeed = false);
+        break;
+    case ReceiverMode::Sync:
+        fiu_do_on(FailPoints::random_receiver_sync_msg_push_failure_failpoint, push_succeed = false);
+        break;
+    case ReceiverMode::Async:
+        fiu_do_on(FailPoints::random_receiver_async_msg_push_failure_failpoint, push_succeed = false);
+        break;
+    default:
+        RUNTIME_ASSERT(false, "Illegal ReceiverMode");
+    }
+}
+} // namespace
+
+template <bool fine_grained_shuffle, bool need_wait>
+std::pair<MPMCQueueResult, ReceivedMessagePtr> ReceivedMessageQueue::pop(size_t stream_id)
+{
+    MPMCQueueResult res;
+    ReceivedMessagePtr recv_msg;
+    if constexpr (fine_grained_shuffle)
+    {
+        if constexpr (need_wait)
+        {
+            res = msg_channels_for_fine_grained_shuffle[stream_id]->pop(recv_msg);
+        }
+        else
+        {
+            res = msg_channels_for_fine_grained_shuffle[stream_id]->tryPop(recv_msg);
+        }
+        if (recv_msg != nullptr)
+        {
+            if (recv_msg->remaining_consumers->fetch_sub(1) == 1)
+            {
+                ReceivedMessagePtr original_msg;
+                auto pop_result [[maybe_unused]] = grpc_recv_queue->tryPop(original_msg);
+                /// if there is no remaining consumer, then pop it from original queue, the message must stay in the queue before the pop
+                /// so even use tryPop, the result must not be empty
+                assert(pop_result != MPMCQueueResult::EMPTY);
+                if (original_msg != nullptr)
+                    RUNTIME_CHECK_MSG(*original_msg->remaining_consumers == 0, "Fine grained receiver pop a message that is not full consumed, remaining consumer: {}", *original_msg->remaining_consumers);
+            }
+        }
+    }
+    else
+    {
+        if constexpr (need_wait)
+        {
+            res = grpc_recv_queue->pop(recv_msg);
+        }
+        else
+        {
+            res = grpc_recv_queue->tryPop(recv_msg);
+        }
+    }
+    return {res, recv_msg};
+}
+
+template <bool enable_fine_grained_shuffle, bool is_force>
+bool ReceivedMessageQueue::pushToMessageChannel(ReceivedMessagePtr & received_message, ReceiverMode mode)
+{
+    std::function<MPMCQueueResult(ReceivedMessagePtr &)> write_func;
+    if constexpr (is_force)
+        write_func = [&](ReceivedMessagePtr & recv_msg) {
+            return msg_channel->forcePush(recv_msg);
+        };
+    else
+        write_func = [&](ReceivedMessagePtr & recv_msg) {
+            return msg_channel->push(recv_msg);
+        };
+    bool success = write_func(received_message) == MPMCQueueResult::OK;
+
+    injectFailPointReceiverPushFail(success, mode);
+    return success;
+}
+
+template <bool enable_fine_grained_shuffle>
```

Comment:
Why we need `enable_fine_grained_shuffle` here? It seems that it's not used.

Reply:
Yes, you are right, after move fine grained write to LooseBoundedQueue, `enable_fine_grained_shuffle ` is not needed.